src/peergos/shared/user/fs/FileWrapper.java

```diff
@@ -1245,43 +1257,61 @@ public SigningPrivateKeyAndPublicHash signingPair() {
-                                    newMapKey, network, crypto)
-                            .thenCompose(updatedBase -> {
-                                return target.addLinkTo(updatedBase, committer, getName(), newCap, network, crypto);
-                            });
-                } else {
-                    return version.withWriter(owner(), writer(), network).thenCompose(snapshot ->
-                            getInputStream(snapshot.get(writer()).props, network, crypto, x -> {})
-                                    .thenCompose(stream -> target.uploadFileSection(snapshot, committer,
-                                            getName(), stream, false, 0, getSize(),
-                                            Optional.empty(), false, network, crypto, x -> {},
-                                            random.randomBytes(32))));
-                }
-            });
-        }).thenApply(newAccess -> true);
+        return context.network.synchronizer.applyComplexUpdate(target.owner(), target.signingPair(),
+                (version, committer) -> version.withWriter(owner(), writer(), network)
+                        .thenCompose(both -> copyTo(target, network, crypto, both, committer)))
+                .thenApply(newAccess -> true);
     }
 
+    public CompletableFuture<Snapshot> copyTo(FileWrapper target,
+                                              NetworkAccess network,
+                                              Crypto crypto,
+                                              Snapshot version,
+                                              Committer committer) {
+        if (! target.isDirectory()) {
+            return Futures.errored(new IllegalStateException("CopyTo target " + target + " must be a directory"));
+        }
+
+        return target.hasChildWithName(version, getFileProperties().name, crypto.hasher, network).thenCompose(childExists -> {
+            if (childExists) {
+                CompletableFuture<Snapshot> error = new CompletableFuture<>();
+                error.completeExceptionally(new IllegalStateException("CopyTo target " + target + " already has child with name " + getFileProperties().name));
+                return error;
+            }
+            if (isDirectory()) {
+                byte[] newMapKey = crypto.random.randomBytes(32);
+                SymmetricKey newBaseR = SymmetricKey.random();
+                SymmetricKey newBaseW = SymmetricKey.random();
+                WritableAbsoluteCapability newCap = ((WritableAbsoluteCapability)target.getPointer().capability)
+                        .withMapKey(newMapKey)
+                        .withBaseKey(newBaseR)
+                        .withBaseWriteKey(newBaseW);
+                return getChildren(version, crypto.hasher, network).thenCompose(children -> {
+                    return target.mkdir(getName(), Optional.of(newBaseR), Optional.of(newBaseW), Optional.of(newMapKey),
```

Comment:
return keyword not required, use simple closure form

Reply:
done